devito/types/dense.py

```diff
@@ -951,6 +936,23 @@ def __init_finalize__(self, *args, **kwargs):
         # Dynamically add derivative short-cuts
         self._fd = generate_fd_shortcuts(self)
 
+        # Flag wether it is a parameter or a variable.
+        # Used at operator evaluation to evaluate the Function at the
+        # variable location (i.e. if the variable is staggered in x the
+        # parameter has to be computed at x + hx/2)
+        self._is_parameter = kwargs.get('parameter', False)
```

Comment:
So I also need to run some interactive sessions to get a better idea of what's going on (and hence may be misrepresenting a little here).
1) So I get that it makes things easier with lazy evaluation but this is something that could (and maybe should) be determined during operator creation (top analysis), not a user defined flag. Is there any way around the user needing to define it?
2) What happens in the following scenario:
User has two sets of equations requiring the consecutive application of two operators. In set one `f` is a parameter, in set two `f` is a variable. Again, maybe misinterpreting how these things work but would seem the user would have two define a 'new' `f` and copy over the data to apply `op2` correctly?

Reply:
1. Don't think it can be inferred in a safe way without ending up either missing some or finding too much
2. I can add a setter to modify it for different operators for that case. I didn't think of that case I I don't have an example in mind where this would happened but that's a good point.

Reply:
2. @FabioLuporini is going to love that :p

Reply:
Haha yeah right :D, I mean it's a purely symbolic property about what to do with it, it doesn't actually change the object in anyway, just what symbolic expressions to generate from it, so I don't think it's that bad.

Reply:
this discussion is scary; I need to take a look at this. Nice that we have a unit test for this now, which should make my life easier.

p2. Not an option

p1. > be determined during operator creation (topo analysis), not a user defined flag

let's talk about this

Reply:
p1 -> not an option., you can turn it however you want  to make it pretty. there is no way you can decide wether an object is a parameter or not.

p2 -> That's a dogmatic answer that does not justify it.

So nothing left to discuss, that's how it is.

Reply:
sometimes you confuse prettiness with engineering flaws. I don't know whether this is gonna be the case. As I said, I still need to go through this